jcache/commons/src/main/java/org/infinispan/jcache/AbstractJCacheNotifier.java

```diff
@@ -76,44 +85,129 @@ public void removeListener(CacheEntryListenerConfiguration<K, V> reg,
          jcache.removeListener(listenerAdapter);
    }
 
+   public void addSyncNotificationLatch(Cache<K, V> cache, K key, V value, CountDownLatch latch) {
+      EventSource<K, V> eventSourceKey = new EventSource<K, V>(cache, key, value);
+
+      Queue<CountDownLatch> latches = new ConcurrentLinkedQueue<>();
+      Queue<CountDownLatch> prev = latchesByEventSource.putIfAbsent(eventSourceKey, latches);
+      if (prev != null) {
+         latches = prev;
+      }
+      latches.add(latch);
+   }
+
+   public void removeSyncNotificationLatch(Cache<K, V> cache, K key, V value, CountDownLatch latch) {
+      EventSource<K, V> eventSourceKey = new EventSource<K, V>(cache, key, value);
+
+      Queue<CountDownLatch> latches = latchesByEventSource.get(eventSourceKey);
+
+      if (latches == null) {
+         return;
+      }
+      latches.remove(latch);
+      if (latches.isEmpty()) {
+         latchesByEventSource.remove(eventSourceKey, EMPTY_QUEUE);
+      }
+   }
+
+   private void notifySync(Cache<K, V> cache, K key, V value) {
+      EventSource<K, V> eventSourceKey = new EventSource<K, V>(cache, key, value);
+
+      notifySync(latchesByEventSource.get(eventSourceKey));
+   }
+
+   private void notifySync(Queue<CountDownLatch> latches) {
+      if (latches == null) {
+         return;
+      }
+      while (true) {
```

Comment:
This was the method I was talking about where concurrent writes for the same key, value would notify more than 1 caller at the same time.  I wonder if it is okay if we just notify 1 latch and return?